src/main/java/com/conveyal/datatools/manager/models/transform/PreserveCustomFieldsTransformation.java

```diff
@@ -0,0 +1,134 @@
+package com.conveyal.datatools.manager.models.transform;
+
+import com.conveyal.datatools.common.status.MonitorableJob;
+import com.conveyal.datatools.manager.models.TableTransformResult;
+import com.conveyal.datatools.manager.models.TransformType;
+import org.supercsv.io.CsvMapReader;
+import com.conveyal.gtfs.loader.Table;
+
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.File;
+import java.io.FileWriter;
+import java.io.StringReader;
+import java.io.IOException;
+import java.nio.file.*;
+import java.util.*;
+import java.util.stream.Collectors;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.ArrayUtils;
+import org.supercsv.io.CsvMapWriter;
+import org.supercsv.prefs.CsvPreference;
+
+
+/**
+ * This feed transformation will attempt to preserve any custom fields from an entered csv in the final GTFS output.
+ */
+public class PreserveCustomFieldsTransformation extends ZipTransformation {
+    /** no-arg constructor for de/serialization */
+    public PreserveCustomFieldsTransformation() {}
+
+    public static PreserveCustomFieldsTransformation create(String csvData, String table) {
+        PreserveCustomFieldsTransformation transformation = new PreserveCustomFieldsTransformation();
+        transformation.csvData = csvData;
+        transformation.table = table;
+        return transformation;
+    }
+
+    @Override
+    public void validateParameters(MonitorableJob.Status status) {
+        if (csvData == null) {
+            status.fail("CSV data must not be null (delete table not yet supported)");
+        }
+    }
+
+    /**
+     * This method creates a hash map of the GTFS table keys to the custom CSV values for efficient lookup of custom values.
+     * The hash map key is the key values of the GTFS table (e.g. stop_id for stops) concatenated by an underscore.
+     * The hash map value is the CsvMapReader (mapping of column to row value).
+     */
+    private static HashMap<String, Map<String, String>> createCsvHashMap(CsvMapReader reader, String[] headers, List<String> tablePrimaryKeys) throws IOException {
+        HashMap<String, Map<String, String>> lookup = new HashMap<>();
+        Map<String, String> nextLine;
+        while ((nextLine = reader.read(headers)) != null) {
+            List<String> customCsvKeyValues = tablePrimaryKeys.stream().map(nextLine::get).collect(Collectors.toList());
+            String hashKey = StringUtils.join(customCsvKeyValues, "_");
+            lookup.put(hashKey, nextLine);
```

Comment:
Since we're doing an action on every single item in the stream, why not do the `lookup.put` from within the stream instead of collecting it into a collection?

Reply:
Well, the lookup relies on the whole output of the stream concatenated with an underscore so we can't really do that inside.